source/core/shape.cc

```diff
@@ -363,6 +363,195 @@ Sphere<dim>::set_position(const Point<dim> &position)
 }
 
 
+template <int dim>
+void
+Superquadric<dim>::closest_surface_point(
+  const Point<dim> &p,
+  Point<dim> &      closest_point,
+  const typename DoFHandler<dim>::active_cell_iterator & /*cell_guess*/)
+{
+  auto point_in_string = this->point_to_string(p);
+  auto iterator        = this->closest_point_cache.find(point_in_string);
+  if (iterator == this->closest_point_cache.end())
+    {
+      this->closest_surface_point(p, closest_point);
+
+      Point<dim> copy_closest_point{};
+      for (unsigned int d = 0; d < dim; d++)
+        copy_closest_point[d] = closest_point[d];
+      this->closest_point_cache[point_in_string] = copy_closest_point;
+    }
+  else
+    closest_point = iterator->second;
+}
+
+template <int dim>
+void
+Superquadric<dim>::closest_surface_point(const Point<dim> &p,
+                                         Point<dim> &      closest_point) const
+{
+  auto point_in_string = this->point_to_string(p);
+  auto iterator        = this->closest_point_cache.find(point_in_string);
+  if (iterator == this->closest_point_cache.end())
+    {
+      // We align and center the evaluation point according to the shape
+      // referential.
+      Point<dim> centered_point = this->align_and_center(p);
+
+      Point<dim>   current_point = centered_point;
+      unsigned int iteration     = 0;
+      unsigned int iteration_max = 1e4;
+
+      Point<dim> dx{}, distance_gradient{};
+      double     current_distance = superquadric(current_point);
+
+      // We scale the relaxation with the maximum blockiness, because higher
+      // blockiness makes convergence more difficult for the gradient descent.
+      const double shape_blockiness =
+        std::max(std::max(exponents[0], exponents[1]), exponents[2]);
+      double relaxation = 0.5 / shape_blockiness;
```

Comment:
Why the default is 0.5? Would it be better if it's 2 in which case for the ellipsoid it doesn't relax since it's a Euclidean distance?

Reply:
Having some relaxation allows to find the "right" surface point. With relaxation = 1, it can happen that we find the surface point on the other side of the shape, especially for narrow concave shapes.
![image](https://github.com/lethe-cfd/lethe/assets/14217956/21c02cbf-4762-4e02-acff-81cbf59ff956)